### PR TITLE
Trim leading whitespace from HTML output #7884

### LINKED ISSUE #7884: Trim leading whitespace from HTML output
Viewing the source of any TEAMMATES page reveals lots of unnecessary whitespace in the HTML:
screen shot 2017-07-31 at 3 35 35 pm

~~This is due to every directive in our `jsp` and `tag` files taking up a newline each. To fix, add the `trim-directive-whitespaces` directive to `src/main/webapp/WEB-INF/web.xml` as detailed here: https://stackoverflow.com/a/20100278~~ Edit: this directive seems to be ignored by Google's runtime.

### PR DESCRIPTION
**Pull request by ongspxm**
Fixes #7884 

**Outline of Solution**
processing of `*.jsp` and `*.tag` files from `src/main/webapp`

### CODE REVIEW

**Comment by wkurniawan07** [comment]
@ongspxm I don't think changing all those whitespaces to `&nbsp;` are necessary, if even a single one is necessary.

**Comment by ongspxm** [comment]
So only spaces between two consecutive directives are stripped?

**Comment by wkurniawan07** [comment]
@ongspxm yes; all others never posed any problem in the first place.

**Comment by ongspxm** [comment]
Roger that, i will rollback the nbsp chnages

**Comment by ongspxm** [comment]
Done and ready for review

**Comment by wkurniawan07** [comment]
@ongspxm the build is not passing. You need to fix the failing test cases.

**Comment by ongspxm** [comment]
looks like most of the manual changes are still required, let me change them to use the `${" "}` style then push a new commit

**Comment by wkurniawan07** [comment]
@ongspxm I gave the Travis build a quick look; the `(Team) :` -> `(Team):` is actually an acceptable kind of change so you may want to modify the test files instead, although keeping the space is an acceptable option too. Other than that, there are still other stuffs that need fix; it is very possible that [this fix](https://github.com/TEAMMATES/teammates/commit/d27d239f15cdd48cd1e38a52ba339c50680b4cc6) is no longer sufficient because many things would have changed since that commit was made.

**Comment by ongspxm** [comment]
yep, i am trawling through the different cases and making the changes

**Comment by ongspxm** [comment]
@wkurniawan07 just to double check, modifying the test files, i run `./gradlew ciTests` with `godMode` on right?

**Comment by wkurniawan07** [comment]
@ongspxm no; Godmode is supposed to be run only locally.

**Comment by whipermr5** [comment]
@ongspxm I think you accidentally committed your bash scripts

**Comment by ongspxm** [comment]
@whipermr5 oops, i will add another commit to remove them

**Comment by ongspxm** [comment]
@wkurniawan07 so when you mention modify the test files, do you mean do it manually for the files in `src/test/resources`?

**Comment by wkurniawan07** [comment]
@ongspxm I realized I misinterpreted your comment; I read it as running that command in Travis with Godmode on, which is not supposed to be the case. Locally, it is up to your discretion whether you want to run `ciTests` or just any individual test with Godmode on, although I'd advise to just isolate the failing cases. You can in fact get the XML file containing the failing test cases from the Travis build log.

Also, I have glanced through your changes. Are you sure all of them are needed? This workaround of `${" "}` comes at the price of readability as @whipermr5 mentioned, so don't use it unless absolutely necessary, and looking at the test cases that actually fail, I'm quite sure that you should not need to add a lot on top of what you have done in 577e9e5.

**Comment by ongspxm** [comment]
According to the output all the spaces surrounding the tags seemed to be stripped, the commit doesnt really cover all the cases.

I trawled through the different cases, use of the tags generally falls into the few cases below.

Cases affecting functionality
- tags generating element properties
- ‎tags generating class names

Cases affecting readability of site
- tags in paragraph (eg. tags generating words like `team` vs `student`)
- tags in ‎formatted text (eg. `Sent: ${numOfEmails}`)

Just to confirm, i only leave the last case untouched?

File: src/main/webapp/WEB-INF/tags/instructor/feedbacks/feedbackSessionPublishButton.tag [hunk 1]
```diff
@@ -1,16 +1,17 @@
+<%@ tag trimDirectiveWhitespaces="true" %>
 <%@ tag description="instructorFeedbacks and instructorHome - Feedback Session publish/unpublish button" pageEncoding="UTF-8" %>
 <%@ taglib uri="http://java.sun.com/jsp/jstl/core" prefix="c" %>
 <%@ attribute name="publishButton" type="teammates.ui.template.FeedbackSessionPublishButton" required="true" %>
 <%@ attribute name="showTooltip" type="java.lang.Boolean" required="true" %>
 <%@ attribute name="buttonType" %>
 <a href="${publishButton.actionAllowed ? publishButton.actionLink : 'javascript:;'}"
-    class="<c:if test="${not empty buttonType}">${buttonType} </c:if>session-${publishButton.actionNameLowercase}-for-test<c:if test="${not publishButton.actionAllowed}"> disabled</c:if>"
+    class="<c:if test="${not empty buttonType}">${buttonType}${" "}</c:if>session-${publishButton.actionNameLowercase}-for-test<c:if test="${not publishButton.actionAllowed}"> disabled</c:if>"
```
[the remark this answers is not in the document]
Comment (ongspxm) [reply src/main/webapp/WEB-INF/tags/instructor/feedbacks/feedbackSessionPublishButton.tag]: roger that, I will try again. thanks for taking the time to guide.

**Comment by wkurniawan07** [comment]
@ongspxm could very well be an adjacent directive causing it. You should try fixing what you can fix first.

**Comment by ongspxm** [comment]
it took awhile, but its finally done

**Comment by whipermr5** [comment]
@ongspxm You accidentally committed the scripts

**Comment by ongspxm** [comment]
sorry, a question, how to rerun a build?

**Comment by whipermr5** [comment]
@ongspxm Close and re-open the PR.

**Comment by ongspxm** [comment]
eh, help? i guess the test stuck at building the environment?

**Comment by ongspxm** [comment]
i guess it is finally rebuilt, but there is some error `The command "curl -fsS -o $HOME/google-cloud-sdk.tar.gz https://dl.google.com/dl/cloudsdk/channels/rapid/google-cloud-sdk.tar.gz" failed and exited with 22 during .`

File: src/main/webapp/WEB-INF/tags/errorPageEmailCompose.tag [hunk 2]
```diff
@@ -6,20 +7,20 @@
   <div class="form-group">
     <label for="error-feedback-email-composer-recipient-email">To:</label>
     <br><em>Note: This is the TEAMMATES support email. It has been pre-filled for your convenience and is not editable. </em>
-    <input type="email" value=<%= Config.SUPPORT_EMAIL %> id="error-feedback-email-composer-recipient-email" class="form-control" name=<%= Const.ParamsNames.ERROR_FEEDBACK_EMAIL_RECEIVER_ADDRESS %> readonly="">
+    <input type="email" value="<%= Config.SUPPORT_EMAIL %>" id="error-feedback-email-composer-recipient-email" class="form-control" name="<%= Const.ParamsNames.ERROR_FEEDBACK_EMAIL_RECEIVER_ADDRESS %>" readonly="">
```
Comment (wkurniawan07) [inline src/main/webapp/WEB-INF/tags/errorPageEmailCompose.tag]: Next time, try to avoid adding in unrelated changes like this, no matter how beneficial they are.

File: src/test/java/teammates/test/cases/datatransfer/AccountAttributesTest.java [hunk 3]
```diff
@@ -115,7 +115,7 @@ public void testGetIdentificationString() {
     @Test
     public void testSanitizeForSaving() {
         AccountAttributes actualAccount = createAccountAttributesToSanitize();
-        AccountAttributes expectedAccount = createAccountAttributesToSanitize();
+        AccountAttributes expectedAccount = actualAccount.getCopy();
```
Comment (whipermr5) [inline src/test/java/teammates/test/cases/datatransfer/AccountAttributesTest.java]: Unrelated change?

Comment (ongspxm) [reply src/test/java/teammates/test/cases/datatransfer/AccountAttributesTest.java]: on one of the ci runs, there was an error where the `expectedAccount` and `actualAccount` differs by the `modifiedDate`.

I traced the creation of the attribute back and since they are supposed to be similar object, this seems to work.

If unrelated, whats the procedue to submit this minor edit?

Comment (whipermr5) [reply src/test/java/teammates/test/cases/datatransfer/AccountAttributesTest.java]: The procedure is to open a new issue for it. However in this case I believe it has already been solved: #8709

Comment (whipermr5) [reply src/test/java/teammates/test/cases/datatransfer/AccountAttributesTest.java]: Ready to merge after you revert this :)

**Comment by ongspxm** [comment]
changes reverted
